enhancements/testing/openshift-tests-extension.md

```diff
@@ -0,0 +1,646 @@
+        "platform": "aws",
+        "architecture": "amd64",
+        # ...others...
+        
+        # Configuration information is also included. 
+        "configuration": {
+            # The configuration id applied to the component
+            # before the test was run.
+            "component": "default",
+        }
+    },
+    
+    # Information about how the test was run
+    "context": {
+        # A small number of seeds can be used to randomize test
+        # parallelization and bucketing. Analyzing seeds against
+        # one another should highlight parallelization issues.
+        # Using the same seed with the same tests should result
+        # in identical planning and parallelization. 
+        "seed": 15,
+        
+        # A sha256 of the tests IDs which were run. This, plus
+        # seed should ensure identical execution pattern between
+        # two runs.
+        "testHash": "deadbeef",
+        
+        # In the future, we may run tests several times against
+        # the same cluster in order to aggregate
+        # test results more quickly. We may want to 
+        # analyze results including this dimension if, for example,
+        # we want to check whether the first run is as consistent
+        # as the following N.
+        "run": 0
+    }
+}
+```
+
+### Risks and Mitigations
+
+#### Binary Incompatibility
+Copying a binary from a different image into a running pod exposes us to binary incompatibility. 
+
+##### CPU Architecture 
+An arm64 binary will not run in an amd64 based pod. `openshift-tests` will need to run on 
+a build farm node with a CPU architecture consistent with architecture of the cluster 
+payload under test.
+
+#### Runtime Size / Speed
+Statically linked binaries are large relative to dynamically linked binaries. As the number
+of extension binaries increase, so will the required tmpfs space used by pods running the
+tests as those binaries are stored locally within the pod.
+
+To mitigate this risk, we need to be aware of it, overprovision test pods with respect to 
+the memory normal test execution requires, and also include a test failure that will 
+TRT attention if the pod is using > 80% of allocatable memory.
+
+#### Image Size
+Statically linked binaries are relatively large. As the number of test extension binaries
+increases, it may lead to a noticeable increase in overall payload size.
+
+#### Poor Extension Implementation
+Providing a contract that must be implemented without by test extensions without centralized 
+control & review of the implementation leaves space for contributions that violate the 
+expectations of that contract.
+
+### Version Skew Strategy
+
+The majority of test extension binaries will reside within the OpenShift payload under test, and thus cannot
+skew. For non-payload components, test filtering, performed by the extension binary itself can take the current
+platform version into account when determining appropriate tests to run.

```

Comment:
One problem that could arise is the openshift-tests binary being used could be newer or older than the payload being run on the cluster under test.  We might want to try to detect this and warn/error.

One implication: imagine a cluster running 4.17 and then using a 4.18 openshift-tests that expects to find an external test binary in the "ovn-kubernetes" image, and it doesn't.

Reply:
Good call - We should code for that implication. If openshift-tests and the cluster under test differ in minor version, failure to extract an external binary from the registry should probably turn from an error to a warning.